engine/grid.py

```diff
@@ -301,79 +320,379 @@ def find_border_nodes(self):
+                            leftmost_node = node
+                            leftmost_node_pos = (row, col)
+
+            self.border_nodes = border_list
+            self.leftmost_node = leftmost_node
+            self.leftmost_node_pos = leftmost_node_pos
+        else:
+            border_list = []
+            leftmost_node = None
+            leftmost_node_pos = None
+
+            for row in range(self.num_rows):
+                for col in range(self.num_cols):
+                    node = self.nodes[row][col]
+                    if node.is_active and self.is_on_border(row, col, self.num_rows, self.num_cols):
+                        # check if this is an active node and on the border
+                        self.nodes[row][col].is_border = True
+                        border_list.append((node, row, col))
+                        if leftmost_node_pos is None or row < leftmost_node_pos[0]:
+                            leftmost_node = node
+                            leftmost_node_pos = (row, col)
+
+            self.border_nodes = border_list
+            self.leftmost_node = leftmost_node
+            self.leftmost_node_pos = leftmost_node_pos
+
+ # --------------------- HELPER FUNCTIONS FOR TRAVERSAL CREATION -------------- #
+    def edge_column_of_next_row(self, pos):
+        """
+        If the direction we are heading is Left, returns the column of the innermost 
+        (closer to the center of the shape) leftmost node that is 
+            1) active  
+            2) not border node 
+        between the current row in [pos] and the next row above [pos].
+
+        If the direction we are heading is Right, returns the column of the innermost 
+        (closer to the center of the shape) rightmost node that is 
+            1) active  
+            2) not border node 
+        between the current row in [pos] and the next row above [pos].
+
+        If the direction we are heading is Up, returns the row of the innermost 
+        (closer to the center of the shape) top node that is 
+            1) active  
+            2) not border node 
+        between the current col in [pos] and the next col above [pos].
+
+        If the direction we are heading is Down, returns the row of the innermost 
+        (closer to the center of the shape) bottom node that is 
+            1) active  
+            2) not border node 
+        between the current col in [pos] and the next col above [pos].
+
+        If there are no more nodes in the next row, return None.
+
+        pos: the current position (col,row)
+        """
+        # Obtain all border nodes in the next and current rows 
+        if self.direction == self.Direction.RIGHT or self.direction == self.Direction.LEFT:
+            # Note: node_info: (node, row, col)
+            #Obtain all border_nodes on the next col
+            
+            candidate_nodes_next = [node_info for node_info in self.border_nodes if node_info[2] == pos[1] + 1]
```

Comment:
maybe a comment about what candidate_nodes_next is? same with candidate_nodes_curr